### PR TITLE
Fix invalid regex quantifier check to not include alternation

### LINKED ISSUE #11650: [BUG] Regex capturing group doesn't work properly in cudf
I tried to filter all numbers in a string (defined by `\s\d+\s`) with cudf but the function `Series.str.replace` does not work in my case while similar expressions in `re/pandas` libraries work properly. Is it a bug?

Here I attach my example:

```python
import re
import pandas as pd
import cudf

print(f"re version:     {re.__version__}")
print(f"pandas version: {pd.__version__}")
print(f"cudf version:   {cudf.__version__}")

text = ["16  6  brr  232323  1  hello  90"]
res = re.sub(r"(^|\s)\d+(\s|$)", " ", text[0])
print(res)

ser = pd.Series(text)
res = ser.str.replace(r"(^|\s)\d+(\s|$)", " ", regex=True)
print(res)

ser = cudf.Series(text)
res = ser.str.replace(r"(^|\s)\d+(\s|$)", " ", regex=True)
print(res)
```
Output:
```
re version:     2.2.1
pandas version: 1.3.1
cudf version:   21.10.01
   brr    hello  
0       brr    hello  
dtype: object
Traceback (most recent call last):
  File "check.py", line 111, in <module>
    res = ser.str.replace(r"(^|\s)\d+(\s|$)", " ", regex=True)
  File "/trinity/home/m.valialshchikov/.conda/envs/lama_venv/lib/python3.8/site-packages/cudf/core/column/string.py", line 907, in replace
    libstrings.replace_re(
  File "cudf/_lib/strings/replace_re.pyx", line 41, in cudf._lib.strings.replace_re.replace_re
RuntimeError: cuDF failure at: ../src/strings/regex/regcomp.cpp:397: invalid regex pattern: nothing to repeat at position 2
```

I believe there is a more elegant solution to my problem (e.g., adding whitespace characters to the beginning and to the end of the string so I don't have to use capturing groups `(^|\s)` and `(\s|$)`) but I wanted to do it that way and it's not clear why it doesn't work as expected.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes the check for invalid quantifier pattern to not include the alternation character. This moves the alternation instruction resolution before the quantifier validation checks.
Added a gtest for this pattern as well.

Closes #11650 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/strings/replace_regex_tests.cpp [hunk 1]
```diff
@@ -162,6 +162,17 @@ TEST_F(StringsReplaceRegexTest, WordBoundary)
...
 
+TEST_F(StringsReplaceRegexTest, Alternation)
+{
+  cudf::test::strings_column_wrapper input(
+    {"16  6  brr  232323  1  hello  90", "123 ABC 00 2022", "abé123  4567  89xyz"});
+  auto results = cudf::strings::replace_re(
+    cudf::strings_column_view(input), "(^|\\s)\\d+(\\s|$)", cudf::string_scalar("_"));
+  auto expected =
+    cudf::test::strings_column_wrapper({"__ brr __ hello _", "_ABC_2022", "abé123 _ 89xyz"});
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(*results, expected);
```
Comment (ttnghia) [inline cpp/tests/strings/replace_regex_tests.cpp:173]: Just curious, why multiple spaces around the characters are not replaced completely by `_`?

Comment (davidwendt) [reply cpp/tests/strings/replace_regex_tests.cpp:173]: The `\s` matches a single space not all contiguous spaces.

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/strings/replace_regex_tests.cpp [hunk 2]
```diff
@@ -162,6 +162,17 @@ TEST_F(StringsReplaceRegexTest, WordBoundary)
   CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(*results, expected);
 }
 
+TEST_F(StringsReplaceRegexTest, Alternation)
+{
+  cudf::test::strings_column_wrapper input(
+    {"16  6  brr  232323  1  hello  90", "123 ABC 00 2022", "abé123  4567  89xyz"});
+  auto results = cudf::strings::replace_re(
+    cudf::strings_column_view(input), "(^|\\s)\\d+(\\s|$)", cudf::string_scalar("_"));
```
Comment (ttnghia) [inline cpp/tests/strings/replace_regex_tests.cpp:170]: Just realized that this is testing for `(^|\\s)` only. Can you add another test with `(\\s|^)` too?

Comment (davidwendt) [reply cpp/tests/strings/replace_regex_tests.cpp:170]: Yes, that is a good idea.

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

**Review by bdice — APPROVED** [review]
Looks good to me!

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**